### ISSUE TITLE
VSCode stuck on restart (the 🔄 button)

### ISSUE THREAD

**Issue by JCKodel** [Dart-Code/Dart-Code#5296]
**Describe the bug**
VSCode is stuck on hot reloading (showing the notification) without ever completing.

On emulator close or stop button pressed, the debug bar won't go away (so the debug never detaches or terminates)

The only solution is to close VSCode and restart it.

**To Reproduce**
1. Open a project that uses `build_runner`
2. Run the project with [F5]
3. Type `Ctrl+Shift+B` to get the build menu (I'm not sure this is Flutter's/Dart's plugin or it is there because of "Dart build_ruinner" VSCode extension. Use the build_runner watch: 
<img width="617" alt="image" src="https://github.com/user-attachments/assets/9dac65e4-6cc8-4f65-9f83-712660f8c7bc">

4. Make your code triggers an exception that will bring VSCode to a debug halt (meaning: when it stops on YOUR code on an exception).
5. Fix the issue, saving the changes and clicking on the hot reload button.

Sometimes, not always, the hot reload will never end. When clicking on the stop button, the app is closed on the emulator, but VSCode keep showing the debug bar (and the hot reloading notification is stuck as well).

This happens on Windows or MacOS, either with Android Emulator or iPhone Simulator (but only when there is a terminal running build_runner watch, which opens a new terminal that keep running for ever):
<img width="1555" alt="image" src="https://github.com/user-attachments/assets/84dc67ce-007d-45f0-9b6c-238b1284a3f2">

**Expected behavior**
To hot reload the app and keep working as usual.

**Please complete the following information:**

<details>

<summary><strong>Workspace Environment</strong></summary>

```text
Dart Code extension: 3.98.0
Flutter extension: 3.98.0 (activated)

App: Visual Studio Code
App Host: desktop
Version: mac 1.93.1

Workspace type: Flutter (LSP)
Workspace name: xxx (Workspace)

Dart (3.5.3): /Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk
Flutter (3.24.3): /Users/jckodel/fvm/versions/3.24.3 (iPhone SE (3rd generation) (ios/ios))

```

</details>

<details>

<summary><strong>Output from 'dart info'</strong></summary>

`/Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk/bin/dart info`

If providing this information as part of reporting a bug, please review the information
below to ensure it only contains things you're comfortable posting publicly.

#### General info

- Dart 3.5.3 (stable) (Wed Sep 11 16:22:47 2024 +0000) on "macos_x64"
- on macos / Version 15.0 (Build 24A335)
- locale is en-BR

#### Project info

- sdk constraint: '^3.5.3'
- dependencies: dart_mappable, device_info_plus, firebase_analytics, firebase_auth, firebase_core, firebase_crashlytics, firebase_performance, flutter, flutter_image_compress, flutter_localizations, gap, google_sign_in, image_cropper, image_picker, in_app_purchase, intl, markdown_widget, no_screenshot, open_settings_plus, path, path_provider, phosphor_flutter, powersync, shared_preferences, ulid, vector_graphics, watch_it, webp
- dev_dependencies: build_runner, dart_mappable_builder, flutter_lints, flutter_test, vector_graphics_compiler, very_good_analysis

#### Process info

| Memory |   CPU | Elapsed time | Command line                                                                               |
| -----: | ----: | -----------: | ------------------------------------------------------------------------------------------ |
|  81 MB |  0.0% |        00:35 | dart devtools --machine --allow-embedding --dtd-uri ws:<path>/RVsHIBv1bkh5qaOf             |
| 511 MB |  0.0% |        00:35 | dart language-server --protocol=lsp --client-id=VS-Code --client-version=3.98.0            |
|  80 MB |  0.0% |        00:35 | dart tooling-daemon --machine                                                              |
|  97 MB |  0.0% |        00:35 | flutter_tools.snapshot daemon                                                              |
|  82 MB |  0.0% |        00:20 | flutter_tools.snapshot debug_adapter                                                       |
| 130 MB | 13.7% |        00:18 | flutter_tools.snapshot run --machine --start-paused -d A657367E-397F-431A-A705-F58BD7F3DAD1 --devtools-server-address http:<path>/ --target <path>/main.dart --web-port 5858 |

</details>

<details>

<summary><strong>Output from 'flutter doctor'</strong></summary>

`/Users/jckodel/fvm/versions/3.24.3/bin/flutter doctor -v`

```text
[✓] Flutter (Channel stable, 3.24.3, on macOS 15.0 24A335 darwin-x64, locale en-BR)
    • Flutter version 3.24.3 on channel stable at /Users/jckodel/fvm/versions/3.24.3
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (3 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/jckodel/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.98.0

[✓] Connected device (4 available)
    • Júlio’s iPhone (mobile)             • xx           • ios            • iOS 18.0 22A3354
    • iPhone SE (3rd generation) (mobile) • xx • ios            • com.apple.CoreSimulator.SimRuntime.iOS-18-0 (simulator)
    • macOS (desktop)                     • macos                                • darwin-x64     • macOS 15.0 24A335 darwin-x64
    • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 129.0.6668.70

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by JCKodel** [comment]
This is the log when the issue is triggered: 
[Dart-Code-Log-2024-09-02 16-37-25.txt.zip](https://github.com/user-attachments/files/17215511/Dart-Code-Log-2024-09-02.16-37-25.txt.zip)

I had to trigger some exceptions, fixes and reloading some 3, 4 times before Hot Reload stuck.

This is the current VSCode state:

<img width="275" alt="image" src="https://github.com/user-attachments/assets/4d9102d0-b606-41ef-821a-30e3ac3388c4">

The app is not running on iPhone simulator (the Stop Button click did stopped the app), but now VSCode is stuck on debug mode.

Clicking Stop do nothing now.

While in this mode and hitting stop twice, this is the log:

```
!! ⚠️ PLEASE REVIEW THIS LOG FOR SENSITIVE INFORMATION BEFORE SHARING ⚠️ !!
Dart Code extension: 3.98.0
Flutter extension: 3.98.0 (activated)

App: Visual Studio Code
App Host: desktop
Version: mac 1.93.1

Workspace type: Flutter (LSP)
Workspace name: xxx (Workspace)

Dart (3.5.3): /Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk
Flutter (3.24.3): /Users/jckodel/fvm/versions/3.24.3 (iPhone SE (3rd generation) (ios/ios))


Logging Categories:
    General, AnalyzerTiming, CommandProcesses, DartTest, DartToolingDaemon, DAP, VmService, DevTools, FlutterDaemon, FlutterRun, FlutterTest, WebDaemon

Tue Oct 01 2024 [16:43:55 GMT-0300 (Brasilia Standard Time)] Log file started
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868) [✓] Connected device (4 available)
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)     • Júlio’s iPhone (mobile)             • 00008030-00092DD92168C02E            • ios            • iOS 18.0 22A3354
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)     • iPhone SE (3rd generation) (mobile) • A657367E-397F-431A-A705-F58BD7F3DAD1 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-18-0 (simulator)
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)     • macOS (desktop)                     • macos                                • darwin-x64     • macOS 15.0 24A335 darwin-x64
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)     • Chrome (web)                        • chrome                               • web-javascript • Google Chrome 129.0.6668.70
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868) [✓] Network resources
[4:43:57 PM] [CommandProcesses] [Info] (PROC 21868)     • All expected network resources are available.

• No issues found!
[4:43:58 PM] [CommandProcesses] [Info] (PROC 21868) exited (0, null)
[4:43:58 PM] [CommandProcesses] [Info] (PROC 21868) closed (0, null)
Tue Oct 01 2024 [16:44:00 GMT-0300 (Brasilia Standard Time)] Log file ended
```

Dart is running:

```
17083 ttys001    0:01.42 /bin/zsh -il
17102 ttys001    0:00.01 /bin/zsh -il
17181 ttys001    0:00.00 /bin/zsh -il
17182 ttys001    0:00.01 /bin/zsh -il
17185 ttys001    0:00.15 /Users/jckodel/.cache/gitstatus/gitstatusd-darwin-x86_64 -G v1.5.4 -s -1 -u -1 -d -1 -c -1 -m -1 -v FATAL -t 16
17277 ttys002    0:00.33 /Applications/Xcode.app/Contents/Developer/usr/bin/xcdevice observe --usb
17278 ttys003    0:00.33 /Applications/Xcode.app/Contents/Developer/usr/bin/xcdevice observe --wifi
21446 ttys006    0:01.07 /Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk/bin/dart --disable-dart-dev --packages=/Users/jckodel/fvm/versions/3.24.3/packages/flutter_tools/.dart_tool/package_config.json /Users/jckodel/fvm/versions/
21475 ttys006    0:01.38 /Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk/bin/dart pub --suppress-analytics run build_runner watch
21488 ttys006    0:44.39 /Users/jckodel/fvm/versions/3.24.3/bin/cache/dart-sdk/bin/dart --packages=/Users/jckodel/Desktop/xxx/src/.dart_tool/package_config.json file:///Users/jckodel/Desktop/xxx/sr
22046 ttys007    0:00.44 -zsh
22105 ttys007    0:00.00 -zsh
22106 ttys007    0:00.00 -zsh
22109 ttys007    0:00.01 -zsh
22110 ttys007    0:00.00 -zsh
22114 ttys007    0:00.05 /Users/jckodel/.cache/gitstatus/gitstatusd-darwin-x86_64 -G v1.5.4 -s -1 -u -1 -d -1 -c -1 -m -1 -v FATAL -t 16
22115 ttys007    0:00.05 sleep 5
```

Killing the first dart process kills the other two, but no change in VSCode.

Tried to kill a bunch of process, but VSCode keep showing the debug bar. The only alternative is to close VSCode =\

**Comment by DanTup** [comment]
Do you have a sample project that triggers this? In the logs it looks like there's a Firebase exception just after the reload:

```
<== [Flutter] [   +8 ms] End of UI start paused handler.
<== [Flutter] [        ] Finished waiting on operations.
<== [Flutter] [   +5 ms] [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: [core/no-app] No Firebase App '[DEFAULT]' has been created - call Firebase.initializeApp()
<== [Flutter]            #0      MethodChannelFirebase.app (package:firebase_core_platform_interface/src/method_channel/method_channel_firebase.dart:195:5)
<== [Flutter]            #1      Firebase.app (package:firebase_core/src/firebase.dart:79:41)
<== [Flutter]            #2      FirebaseAuth.instance (package:firebase_auth/src/firebase_auth.dart:38:47)
<== [Flutter]            #3      FirebaseAuthProvider._initialize (package:maismulher/auth/infra/firebase_auth_provider.dart:32:18)
<== [Flutter]            #4      FirebaseAuthProvider.factory (package:maismulher/auth/infra/firebase_auth_provider.dart:18:20)
<== [Flutter]            #5      _GetItImplementation._register.<anonymous closure> (package:get_it/get_it_impl.dart:1256:48)
<== [Flutter]            #6      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:861:45)
<== [Flutter]            #7      Future._propagateToListeners (dart:async/future_impl.dart:890:13)
<== [Flutter]            #8      Future._addListener.<anonymous closure> (dart:async/future_impl.dart:472:9)
<== [Flutter]            #9    <…>
```

It's not clear to me if this is related to the issue, but it might be easier to determine that given a repro.

> When clicking on the stop button, the app is closed on the emulator, but VSCode keep showing the debug bar (and the hot reloading notification is stuck as well).

I've seen this a few times too - it seems like VS Code sometimes gets stuck if the debugger doesn't terminate on the first graceful request. I've never been able to repro this well enough to debug and file a VS Code issue though - if you're able to trigger this reliably, I would also be interested in steps/project to repro that.

**Comment by JCKodel** [comment]
First, I misnamed the problem: it's on hot restarting, not hot reloading.

The attached project triggers this issue.

First, I've added a lib that uses build_runner (dart_mappable, in this case, but it really doesn't matter - the problem is on build_runner watch).

Then, everything worked fine. I could not trigger the issue. One thing I've noticed was: in a bad project, VSCode breaks in debug mode, then the exception is dumped in the debug console AFTER the hot restart, but in a new project, the exception is dumped to console, then VSCode breaks in debug mode.

I was about to quit, but then I've realized: in all those projects, there were always some package that had native code (such as Firebase).

So, I've added `shared_preferences` in the project and trigger an exception while using it (basically, trying to get the length of an null string).

This is what caused VSCode to be stuck.

So, basically:

1) Using https://marketplace.visualstudio.com/items?itemName=Kaiqun.build-runner, type `Ctrl`+`Shift`+`B` to get the build task runner (don't know if this menu is from the VSCode extension above or if it is native), then choose build_runner watch:

<img width="640" alt="image" src="https://github.com/user-attachments/assets/b0fad7c9-846e-418c-849c-5f0e6982ea0f">

2) While build_runner is running inside VSCode terminal, make some changes, make it run (by changing some class that triggers a build_runner rebuild) and make some exception in native code (meaning: calling a package that has native Kotlin/Swift code, such as shared preferences).

Eventually, VSCode will stuck (in my test, was in the first time).

[Archive.zip](https://github.com/user-attachments/files/17246507/Archive.zip)

**Comment by DanTup** [comment]
I don't seem to be able to repro this, however it's not clear to me if I need to make a specific kind of change for:

> While build_runner is running inside VSCode terminal, make some changes, make it run (by changing some class that triggers a build_runner rebuild) and make some exception in native code (meaning: calling a package that has native Kotlin/Swift code, such as shared preferences).

I was just modifying the name of the key in the `prefs.getString()` call. When I save, it triggers both the build_runner code and the hot reload, but since you said it was a hot restart, I'm then manually pressing hot restart.

Is timing a factor here? Are you pressing Hot Restart while the build_runner task is running? There could be some races in doing so (since the Hot Restart is not going to wait for an external task), but it's not clear if that's the problem.

**Comment by JCKodel** [comment]
The issue happens when an exception is thrown, VSCode breaks in your code then, while VSCode is paused on debug, you hit hot restart.

I usually do this as a continue, because Dart throws so many exceptions in a row that continue almost means hit F5 like crazy until the program is flowing again.

Instead, when a exception occurs, I usually fix the problem while VSCode is paused, then I manually hit hot restart to restart the application. <-- That's when the issue occurs and, as far as I could see, only when build_runner is running.

**Comment by DanTup** [comment]
I've managed to reproduce this a few times using the project above, where I:

- tap the FAB to break on exception
- modify the file (without saving)
- click hot restart on the toolbar (which saves)

When this happens, we trigger a hot-reload-on-save, and the hot-restart (from clicking the button), and build runner rewrites some of the files. I feel like there is some race condition between some of these, however I cannot reproduce it while logging is running, or from a script that should be doing exactly the same.

@petrnymsa do you have a simpler repro for this?

**Comment by JCKodel** [comment]
@DanTup Glad you were able to reproduce.

In your reproduction, native code (such as Firebase or SharedPreferences) plays a role or it was just my impression?

**Comment by DanTup** [comment]
@JCKodel it was using your example project, with the exception in shared_preferences code. I'm not yet sure if that's necessary or is just affecting the timing (enabling verbose logging seems to "fix" it, so I think it's very sensitive to the timing of some of the things happening!).

**Comment by JCKodel** [comment]
@DanTup It happened again for me, but, this time, although the project has packages with native code, no code whatsoever was making platform calls.

I was debugging this piece of code, then I've changed it while VSCode was paused on a breakpoint (no exceptions this time), then, after saving, Hot Reloading notification was stuck (this time I didn't hit hot restart button).

```dart
String _getContext() {
  final frames = StackTrace.current.toString().split("\n");
  final frame = frames.elementAtOrNull(2);
  late final String moduleName;
  late final String functionName;

  if (frame != null) {
    // I was messing around this line, addind this replaceFirst:
    final tokens = frame.replaceFirst(".<anonymous closure>", "").split(" ");
    final context = tokens[tokens.length - 2].split(".");

    moduleName = context.first;

    final fn = context.elementAtOrNull(1);

    functionName = fn == null ? "" : ".${fn}";
  } else {
    moduleName = functionName = "";
  }

  return "${moduleName}${functionName}";
}
```

Since you said you think there is some race condition in there, there is something peculiar about my VSCode configuration:

I use Dart Import Sorter (aziznal.dart-import-sorter) and dart-import (luanpotter.dart-import) VSCode extensions. Those extensions are bugged and they change the .dart file after VSCode save them, so they are always marked as not saved.

So I have to turn those options in VSCode settings: `dartImport.fixOnSave": false` and `"dartimportsorter.sortOnSave": false`.

Then, I use the multi-command (ryuta46.multi-command) extension to run this on save (the way I did this was: override the Ctrl+S, only when the file extension is .dart):

My keybinding:

```json
{
  "key": "cmd+s",
  "command": "extension.multiCommand.execute",
  "args": {
    "command": "multiCommand.fixAndSaveDart"
  },
  "when": "editorTextFocus && !editorReadonly && resourceExtname == '.dart'"
},
```

My multi-command setting:

```json
"multiCommand.commands": {
  "multiCommand.fixAndSaveDart": {
    "interval": 100,
    "sequence": [
      "dart-import.fix",
      "dartimportsorter.sortImports",
      "workbench.action.files.save"
    ]
  },
  "multiCommand.fixAndSaveAllDart": {
    "interval": 100,
    "sequence": [
      "dart-import.fix",
      "dartimportsorter.sortImports",
      "workbench.action.files.saveAll"
    ]
  }
},
```

Even so, in some rare events that I don't know how are triggered, the file is not saved (I mean, it is saved, but the changes are made after that, so they are left unsaved). Hitting Ctrl+S again solves the issue.

Maybe there is some hiccup in this process that is triggering this issue a lot more often for me.

**Comment by JCKodel** [comment]
> Bump so the annoying bot doesn't close it.

**Comment by JCKodel** [comment]
The problem has worsened since the opening of this issue.

Now, *anytime* a change is made in code while VSCode is paused on debug (either by exception, `debugger()` or break-point) and the hot restart buttom is clicked, the bug manifests (hot restart never completes, then stop never stops, then VSCode must be restarted).

In other words: it is not only exceptions.

**Comment by JCKodel** [comment]
No, I cannot =\

My actual project fails every single time, but I could not replicate it on a new project. I thought it was some long-running isolates (such as SQflite and FFIBlurHash), but running a long-running isolate with `integral_isolates` package didn't work. Also, I've added the `sqflite_common_ffi`, which uses isolates, even did a query to ensure that, and everything works just fine.

I understand how hard is to track this down given this results =\

**Comment by JCKodel** [comment]
For me always occurred in iPhone simulator.

After doing some refactoring in my code (removing powersync & sqlitef and using drift), it never occurred again. Also, I did disable some extensions (I can only think of `Dart build_runner` having something to do with it, since my first contact with this bug seems to always be related with `dart run build_runner watch`, but it could be only a coincidence, since I cannot pinpoint anything).

There were an open issue in powersync repository, and they said sometime a database connection were not closed correctly on hot restart and some error occurred within powersync. Maybe, there is something that some isolate or "native" library does that freaks up VSCode, such as not closing something or expecting something that no longer exists?

@vasilich6107 Can you say something about: 1) what extensions you use, 2) what Flutter packages you use that uses isolates or native code. Maybe we have some common usage that can help to pinpoint what causes it.

**Comment by DanTup** [comment]
Sorry for the delay.

> Could you give some steps how to properly capture the logs to help with debug?

Running the **Dart: Capture Debugging Logs** command captures the logs here, although reliable steps to reproduce are much more useful than logs. Unfortunately with @JCKodel's repro I wasn't able to reproduce while capturing logs which made things more difficult to understand.

I'm not that surprised that this occurs in AS/IntelliJ too, as it doesn't feel like something VS Code-specific (generally both IDEs are just ultimately running `flutter run --machine` and sending the same kinds of commands for hot reload). Perhaps since we're certain of that now, it's better to file this at https://github.com/flutter/flutter/ with whatever repros you have between you. Even if it's caused by third party native code, perhaps the Flutter tool could handle this better (and if there's not much it can do, perhaps emitting a warning after a number of seconds indicating something may be wrong and some pointers for troubleshooting?)? 🤔

**Comment by JCKodel** [comment]
@DanTup Previously, I said it was happening anymore, but I was mistaken.

The reason it was not happening was because I've configured VSCode to not break on exceptions, since the [framework](https://pub.dev/packages/streamline) I'm using automatically stops on exceptions (using `debugger()`):

<img width="435" alt="Image" src="https://github.com/user-attachments/assets/e7412c6e-21f3-4390-af98-76b935fe6a95" />

When this `Uncaught Exceptions` is on and some exception makes VSCode to break into code, and then we change something and then hit hot restart, the problem arises.

When my code stops VSCode using the `debugger()` method from this piece of code:

```dart
PlatformDispatcher.instance.onError = (error, stack) {
  if (kReleaseMode) {
    FirebaseCrashlytics.instance.recordError(error, stack, fatal: true);
  } else {
    final message = "PlatformDispatcher thrown ${error.runtimeType}";

    log(message, error: error, stackTrace: stack);
    debugger();
  }

  return true;
};
```

...VSCode never freezes.

**Comment by JCKodel** [comment]
@DanTup, still no.

I tried to create a project using Drift and BlurhashFFI, because they both create a long running isolate:

<img width="445" alt="Image" src="https://github.com/user-attachments/assets/a06fd590-c429-4426-af99-93aa57778105" />

Then I tried to keep build_runner running with watch:

<img width="552" alt="Image" src="https://github.com/user-attachments/assets/af669365-ccec-401f-a33d-5fbc12265731" />

Then I edited the Drift SQL generated code to thrown an exception in SQLite, but everything works just fine.

In 2 real projects, the VSCode freezes every single time I hot restart while the debugger is stopped. =\ Really don't know what can be the cause =\

**Comment by JCKodel** [comment]
@DanTup Maybe this was the reason? https://github.com/flutter/flutter/issues/161466

**Comment by DanTup** [comment]
@JCKodel very possible. There is also https://github.com/flutter/flutter/issues/165656 open which is related to iOS/iproxy.

I think the first one was included in a hotfix to Flutter recently, so if it's that you're hitting, it might not be resolved if you upgrade to latest. Are you able to check this? If you still see it, you could try following the instructions at https://github.com/flutter/flutter/issues/165656#issuecomment-2766546922 to update `iproxy` to see if that solves it.

Thanks!

**Comment by JCKodel** [comment]
@DanTup After doing that, the bug isn't triggered, so, I guess, it is fixed (by the 161466 in conjunction with the tooling update).

**Comment by DanTup** [comment]
Great, thanks for confirming!

**Comment by JCKodel** [comment]
@DanTup I'm seeing this again **EVERYTIME** an exception triggers the debugger. =\ It's even worse than before.

**Comment by DanTup** [comment]
@JCKodel it sems odd for the VS Code version to make a difference here - however there is an issue with VS Code showing incorrect state of threads if events arrive close together - I'm not sure if that might be affecting you:

https://github.com/microsoft/vscode/issues/238591

Did you ensure you're on the latest Flutter (with the hotfix noted above) and try the instructions to upgrade iproxy to see if that makes a difference?

**Comment by JCKodel** [comment]
@DanTup 

>  see if that makes a difference

I've reapplied the fix and still have issues. Any exception freezes VSCode =\

**Comment by JCKodel** [comment]
> [@JCKodel](https://github.com/JCKodel) by "freeze", do you mean locks up the UI?
>

By "freeze" I mean: the notification "Hot restarting..." never closes. If I hit hot restart again, a new notification appears. If I click the STOP button on VSCode debug bar, the app closes, but the debug bar don't close, so at that point I can't start a new debug session nor stop the old one. The only option here is to restart VSCode and hit [F5] again to start debug a new session.


> If so, we should file this with VS Code because we should never be able to freeze anything like that.
> 

VSCode remains operational (I can edit code, etc.) the only thing that doesn't work anymore is Debug (either Run [F5] or stop).

> Can you repro this in a brand new project? Can you share exact versions of VS Code, Dart-Code, Flutter and what host+target platforms you're using so I can try to repro?
> 

No. This happens only when multiple isolates are being run (I've tried multiple times to make this happen in a new project, but it doesn't trigger. The current project is not old (it is in the very beginning), but I can pinpoint what I did to start having problems). 

These are the isolates (from packages blurhash_ffi, drift, drift_sqlite_async, powersync, sqlite_async):


<img width="436" alt="Image" src="https://github.com/user-attachments/assets/69d3983c-a1b1-44f8-8abf-0f74799bcf96" />



> (A log collected with **Dart: Capture Debugging Logs** from the newly created project would also help, so I can compare to mine).
> 

When *I* throw something or `debugger(message: "pause");`, the hot restart works without issues every single time (of course it does >.<). Since I cannot reliable make the bug happens at my will, I'll wait until it happens again, then I post the logs here.

**Comment by JCKodel** [comment]
@DanTup:

Ok, now I can reliable trigger this issue.

First, it seems related to running isolates. These are the threads running on the offending Flutter project:

<img width="387" alt="Image" src="https://github.com/user-attachments/assets/1c02635d-47bb-4a1a-9781-3d19307f0a87" />

The issue occurs when I add some checks in this if:

```dart
void _initializeDebugMode() {
  FlutterError.onError = (FlutterErrorDetails flutterErrorDetails) {
    log(
      "🔴 [${flutterErrorDetails.library}] ${flutterErrorDetails.exception.runtimeType}",
      error: flutterErrorDetails.exception,
      stackTrace: flutterErrorDetails.stack,
    );

    if (flutterErrorDetails.exceptionAsString().contains(
          "A RenderBox object must have an explicit size",
        ) ||
        flutterErrorDetails.exceptionAsString().contains(
          "BoxConstraints forces an infinite height.",
        ) ||
        flutterErrorDetails.exceptionAsString().contains(
          "NEEDS-PAINT NEEDS-COMPOSITING-BITS-UPDATE",
        ) ||
        flutterErrorDetails.exceptionAsString().contains(
          "'child!.hasSize': is not true",
        ) ||
        flutterErrorDetails.exceptionAsString().contains(
          "!childSemantics.renderObject._needsLayout",
        )) {
      return;
    }

    debugger();
  };

  PlatformDispatcher.instance.onError = (error, stack) {
    log(
      "🔴 [PlatformDispatcher] ${error.runtimeType}",
      error: error,
      stackTrace: stack,
    );

    debugger();
    return true;
  };
}
```

I'm doing this because exceptions breakpoints are very annoying (especially on layout issues like this, where I have to hit [F5] Continue dozens of time, so instead I often run my projects with [ ] All Excedptions and [ ] Uncaught Exceptions tuned off).

So, while VSCode is stopped on that `debugger()` line, I copy one of the checks (the `|| contains`), save the file and hit the Restart button.

At that moment, this is the log:

[Dart-Code-Log-2025-04-03 19-09-44.txt.zip](https://github.com/user-attachments/files/20495665/Dart-Code-Log-2025-04-03.19-09-44.txt.zip)

**Comment by Colton127** [comment]
I encountered this thread after Googling a similar problem of mine. Instead of VS Code getting stuck, my entire app would crash when hot reloading. I tracked the problem down to blurhash_ffi, and upon replacing it with another blurhash decoder, the problem resolved. See if replacing blurhash_ffi with another blurhash package resolves your issue. I hope this helps.

**Comment by JCKodel** [comment]
@Colton127 Thanks, but blurhash_ffi works fine. The issue also presents itself without it.

The only thing in common, as far as I know, are long-running isolates.

**Comment by Colton127** [comment]
@JCKodel You're right - the issue has nothing to do with blurhash_ffi, or any plugin for that matter. I created a simple isolate that practically does nothing, and hot reloading while it is running immediately crashed my app with an out-of-memory error. I think our issues are interlinked, but manifest differently. I'll create my own issue and update you on what I find.

Edit: And like you, I cannot replicate the issue in a new project. 

Have you tried opening DevTools in a browser and monitoring memory usage during reloads?

**Comment by DanTup** [comment]
@JCKodel thanks for the log! It doesn't reveal a lot, but it does show that Flutter starts a hot restart, the isolates all exit, and then nothing (until you click Stop).

To confirm - is the issue here with Hot Restart, or Hot Reload? (the log shows a Hot Restart not working, but the title of this issue says reloading, so I'm not sure if the restart was an attempt to fix a reload that was already broken, or the issue is with restart).

One thing that _might_ help, but will produce much larger logs, is to add `-v` to your `args` in `launch.json` (or `dart.flutterRunAdditionalArgs` in your _Workspace_ settings for this project). Of course, these logs will be much more verbose so as always, be sure there's nothing sensitive when sharing.

There was another issue raised recently at https://github.com/Dart-Code/Dart-Code/issues/5517 that may or may not be the same / related.

**Comment by DanTup** [comment]
@bkonyi FYI, in case you have any additional ideas. Something seems to be getting stuck during hot reloads / and or restarts. Are there any tools in DevTools that might be useful when the app gets into this state to understand where/why it is stuck?

**Comment by JCKodel** [comment]
@DanTup It's **restart**. Hot reload never failed. I'm changing the issue title to fix that.

The #5517 happened to me as well 2 days ago. I don't know why, but both the **restart** and **hot reload** was taking from 4 to 8 seconds to be applied.

To be clear, I'm talking about this (I've edited this example to illustrate what I was seeing, the reassemble part was taking huge amount of time):

```
Reloaded 1 of 3764 libraries in 765ms (compile: 19 ms, reload: 516 ms, reassemble: 4183 ms).
```

I didn't change anything dramatic in the code to make this happen, and even simple things, such as changing a text in `Text()` widget took too much time. The next day, everything worked just fine (so, it was something that a machine restart fixed). And, to be clear, we're not talking about lack of resources (i7 13700, 128Gb RAM, nVme 7000gb/s write speed).

What I also noticed is that sometimes the application hangs on the device for some reason (maybe some native code, since I'm using some native extensions, such as sqlite, ffi, etc.) and that made exactly what #5517 shown (hot reload notifications queueing up). But, eventually, things went through (it's the same symptom I described in this issue, with the difference that, after some several seconds, everything works). Maybe this could be some clue.

My apps has nothing special, with the exception of native packages. Every time I talk to you (and thanks for your patience), there is something native involved: shared preferences, sqlite, blurhash_ffi, etc. Those things alone don't trigger the issue, but in a real app, with multiple isolates running, then we get the bug =(

I'll try to get more detailed logs as soon as possible (for now, I'm working with debug off 100% of the time, relying to the console output for exceptions thrown (which, btw, sometimes stop working as well, the debug console just refuses to show ANY information at all, restarting vs solves)).

**Comment by DanTup** [comment]
> To be clear, I'm talking about this (I've edited this example to illustrate what I was seeing, the reassemble part was taking huge amount of time):

So is the issue just the spikes in time, or that is gets stuck and never completes? (The log you provided had a restart that didn't complete, although it was terminated after around 5-10s IIRC, so it's not clear if it might have completed after a while).


> I'll try to get more detailed logs as soon as possible

Thanks!


> btw, sometimes stop working as well, the debug console just refuses to show ANY information at all, restarting vs solves

It's probably worth filing an issue about this too, although without logs (which you'd need to be capturing before it happens) or a repro, it may also be difficult to get to the bottom of. It'd be interesting to know if it's only output that breaks, or all debugging functionality. I also suspect it's not the issue, but FWIW some users have reported the issue of no output before, and it was because the "filter" box in the debug console can be quite subtle, so if you're typed into it (even accidentally) it might not be obvious it's filtering everything out.

**Comment by JCKodel** [comment]
> So is the issue just the spikes in time, or that is gets stuck and never completes? 

It happened only one day, consistently spiking the time, but no other effect. It just took too long to hot reload and refresh.

> although without logs

That's the tricky part. It happens after several hours working (it seems like those situations where there is so much leak in a software that you are forced to restart it). It could be extensions, it could be a zillion things, and keep the log on for 10, 12 hours (my daily workload), it's tricky =(

>  it's only output that breaks

Yes. Only the output stops. Nothing else unusual.

> because the "filter" box in the debug console can be quite subtle

When that happened for Android, I thought it was the filters we must add because Android log is soooo noisy, but this is a relatively fresh MacOS install and the filter box is clear (also, a restart would not fix it since the filters are persistent).

**Comment by JCKodel** [comment]
Don't know if it is useful for this bug, but, again, the hot reload is slow. I'm just messing with a simple function that tries to get the current GoRouter matched location in a stateless build method, *while the debugger is stopped on a breakpoint** (i.e.: nothing fancy and the app state doesn't do anything special so far). There is some threads running:

<img width="383" alt="Image" src="https://github.com/user-attachments/assets/b5120b51-fa6a-432e-b077-dfd6527ebfa5" />

Out of nowhere, the hot reload is slow again:

```
Reloaded 4 of 2909 libraries in 683ms (compile: 34 ms, reload: 446 ms, reassemble: 184 ms).
Reloaded 4 of 2909 libraries in 4,684ms (compile: 16 ms, reload: 419 ms, reassemble: 4234 ms).
Reloaded 1 of 2909 libraries in 4,579ms (compile: 13 ms, reload: 424 ms, reassemble: 4131 ms).
```

After the first slow, I turned on the log and this is it:

[Untitled.log.zip](https://github.com/user-attachments/files/20635706/Untitled.log.zip)

After restarting VSCode and uninstalling the app from the iOS simulator, times are still high (those are the first reports after I just save a file without changes):

```
Reloaded 1 of 2909 libraries in 2,606ms (compile: 17 ms, reload: 429 ms, reassemble: 2143 ms).
Reloaded 1 of 2909 libraries in 4,372ms (compile: 11 ms, reload: 428 ms, reassemble: 3919 ms).
Reloaded 1 of 2909 libraries in 4,371ms (compile: 12 ms, reload: 413 ms, reassemble: 3933 ms).
```

Hot restart is quite fast:

```
Restarted application in 486ms.
```

I'm putting timestamps in the main method and in the build method I'm messing with and hot restarting:

```
flutter: 2025-06-06 21:09:48.883252
flutter: 2025-06-06 21:09:49.322957
```

So, my app is wasting 0.5s to render itself (what I mean is: there is nothing blocking the app while hot reloading. It is *NOT* a problem with my code (especially because before the bug triggered, my hot reloadings were all under 190ms)).


EDIT: While removing those print(DateTime.now()), I've noticed that two notifications "Hot reloading..." appears on the screen. VSCode is waiting for something for each changed file.

This last scenario is logged here:

[Dart-Code-Log-2025-05-05 21-12-33.txt.zip](https://github.com/user-attachments/files/20635763/Dart-Code-Log-2025-05-05.21-12-33.txt.zip)

EDIT 2: Even after a machine restart, the hot reload is still slow. Then, I just unchecked the Breakpoints: [ ] Uncaught Exceptions, now the hot reload is fast again:

```
Reloaded 4 of 2909 libraries in 4,463ms (compile: 27 ms, reload: 499 ms, reassemble: 3924 ms).
Reloaded 1 of 2909 libraries in 2,593ms (compile: 11 ms, reload: 490 ms, reassemble: 2082 ms).
Reloaded 1 of 2909 libraries in 3,031ms (compile: 13 ms, reload: 504 ms, reassemble: 2503 ms).
Reloaded 1 of 2909 libraries in 636ms (compile: 11 ms, reload: 448 ms, reassemble: 166 ms).
Reloaded 1 of 2909 libraries in 646ms (compile: 13 ms, reload: 440 ms, reassemble: 175 ms).
Reloaded 1 of 2909 libraries in 657ms (compile: 11 ms, reload: 463 ms, reassemble: 171 ms).
```

**Comment by DanTup** [comment]
@JCKodel sorry for the delay. I'm a little confused about whether we're looking at reloads or restarts here - above you said restarts, but the last log file seems to only contain reloads.

A lot of what happens during a reload/restart is inside Flutter and not something I'm super familiar with. However, there are things in the debug adapter that will add some delays because we must tell Flutter to start the application _paused_ so that we can perform configuration (adding breakpoints to code, setting break-on-exception behaviour etc.) so that would certainly add _some_ delay, but whether it accounts for the time you're seeing, I'm not sure.

Are you able to capture a log file while performance a reload or restart (whichever the issue is), and include the single line of output text for a slow restart that is captured in the log, so I can review what's happening there specifically? Please try to reproduce this with just one isolated reload/restart and not multiple to keep the logs clearer and easier to understand.

(It may also be beneficial to capture this with the `-v` argument in the launch.json, but even with that the log might understand if the time is spend during the paused phase to set up breakpoints etc.)

**Comment by DanTup** [comment]
@JCKodel are you still seeing this? Are you able to respond to the comments above? Thanks!

**Comment by JCKodel** [comment]
> [@JCKodel](https://github.com/JCKodel) are you still seeing this? Are you able to respond to the comments above? Thanks!

I'm not sure. Because of these bugs I'm using VSCode with debugger off most of the time, so I can't tell if it is fixed or not. I'll try to check that.
